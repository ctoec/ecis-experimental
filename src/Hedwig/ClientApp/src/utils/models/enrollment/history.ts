import { ProcessStepProps } from '../../../components/ProcessList/ProcessStep';
import { DeepNonUndefineable, DeepNonUndefineableArray } from '../../types';
import {
	Enrollment,
	Funding,
	FundingSource,
	FamilyDetermination,
	C4KCertificate,
} from '../../../generated';
import { prettyAge, prettyFundingTime, NO_FUNDING } from '..';
import dateFormatter from '../../dateFormatter';
import { dateSorter } from '../../dateSorter';
import moment from 'moment';
import { c4kCertificateSorter } from '../c4kCertificate';
import { fundingStartSorter, getFundingTime } from '../funding';

// sortable process step props include:
// - sortDate, the field to use for sorting
// - sortWeight, a number to use as a sorting tiebreaker to ensure enrollment starts come before funding/determination steps,
// 	and enrollment ends come after
type SortableProcessStepProps = ProcessStepProps & { sortDate: Date | null; sortWeight: number };

/**
 * Generates SortableProcessStepProps array for a given array of enrollments,
 * also creating step props for fundings and family determinations
 * @param enrollment
 * @returns sorted SortableProcessStepProps array
 */
export function getEnrollmentHistoryProps(
	currentEnrollment: DeepNonUndefineable<Enrollment>
): ProcessStepProps[] {
	// allEnrollments is sorted in descending order by Entry
	// (pastEnrollments are sorted thusly in API response)
	var allEnrollments = [
		currentEnrollment,
		...(currentEnrollment.pastEnrollments ? currentEnrollment.pastEnrollments : []),
	];

	var processStepProps: SortableProcessStepProps[] = [];

	allEnrollments.forEach((enrollment, idx) => {
		processStepProps = processStepProps.concat(
			enrollmentStepProps(enrollment, idx, allEnrollments.length)
		);

		// funding steps
		processStepProps = processStepProps.concat(
			fundingStepProps(enrollment.fundings, enrollment.exit)
		);

		// only add determination steps and C4K funding steps for the current enrollment (others will be dupes)
		if (idx === 0) {
			// determination steps
			processStepProps = processStepProps.concat(
				determinationStepProps(enrollment.child.family.determinations)
			);

			// C4K certification steps
			processStepProps = processStepProps.concat(
				c4kCertificateStepProps(enrollment.child.c4KCertificates)
			);
		}
	});

	// return all steps sorted by `sortDate`, and sortWeight
	return processStepProps.sort(enrollmentHistoryPropsSorter);
}

export function enrollmentStepProps(
	_enrollment: DeepNonUndefineable<Enrollment>,
	enrollmentIndex: number,
	allEnrollmentsCount: number
) {
	var processStepProps: SortableProcessStepProps[] = [];

	// enrollment start step
	if (_enrollment.entry) {
		processStepProps.push(
			getSortableStep({
				heading:
					enrollmentIndex === allEnrollmentsCount - 1
						? // heading text for first enrollment (last in ordered array)
						  `Enrolled in ${_enrollment.site.name}`
						: `Changed enrollment to ${prettyAge(_enrollment.ageGroup)} in ${
								_enrollment.site.name
						  }`,
				body: `on ${dateFormatter(_enrollment.entry, false)}`,
				stepDate: _enrollment.entry,
				stepWeight: -1,
			})
		);
	}

	// enrollment stop step (only care about withdrawal for the last enrollment)
	if (enrollmentIndex === 0 && _enrollment.exit) {
		processStepProps.push(
			getSortableStep({
				heading: `Withdrawn from ${_enrollment.site.name}`,
				body: `on ${dateFormatter(_enrollment.exit, false)}`,
				stepDate: _enrollment.exit,
				stepWeight: 1,
			})
		);
	}

	return processStepProps;
}

/**
 * Generates SortableProcessStepProps array for a given array of fundings
 * @param fundings
 * @param enrollmentExit the value of Exit on the owning enrollment,
 * used to determine if switch to private pay step should be created
 */
export function fundingStepProps(
	fundings: DeepNonUndefineable<Funding[]> | null,
	enrollmentExit: Date | null
) {
	var processStepProps: SortableProcessStepProps[] = [];

	if (fundings) {
		fundings.sort(fundingStartSorter).forEach(funding => {
			// CDC Funding start if:
			// - source is CDC
			// - funding has first reporting period
			if (funding.source === FundingSource.CDC && funding.firstReportingPeriod) {
				var heading = `Switched funding to ${funding.source}`;
				var fundingTime = getFundingTime(funding);
				if (fundingTime) heading += ` - ${prettyFundingTime(fundingTime)}`;

				processStepProps.push(
					getSortableStep({
						heading,
						body: `on ${dateFormatter(funding.firstReportingPeriod.periodStart, false)}`,
						stepDate: funding.firstReportingPeriod.periodStart,
					})
				);
			}

			// Switch to private pay if:
			// - source is CDC
			// - funding has last reporting period
			// - enrollment is still ongoing (exit == null)
			// 	 OR last reporting period ends before enrollment exit
			// TODO: Handle other funding sources
			if (
				funding.source === FundingSource.CDC &&
				funding.lastReportingPeriod &&
				(!enrollmentExit || funding.lastReportingPeriod.periodEnd < enrollmentExit)
			) {
				var dayAfterLastReportingPeriodEnd = moment(funding.lastReportingPeriod.periodEnd)
					.add(1, 'days')
					.toDate();
				processStepProps.push(
					getSortableStep({
						heading: `Switched funding to ${NO_FUNDING}`,
						body: `on ${dateFormatter(dayAfterLastReportingPeriodEnd, false)}`,
						stepDate: dayAfterLastReportingPeriodEnd
					})
				);
			}
		});
	}

	return processStepProps;
}

/**
 * Generates SortableProcessStepProps array for a given array of family determinations
 * @param determinations
 */
export function determinationStepProps(
	determinations: DeepNonUndefineable<FamilyDetermination[]> | null
) {
	var processStepProps: SortableProcessStepProps[] = [];
	if (determinations) {
		determinations.forEach(determination => {
			if (!determination.notDisclosed && determination.determinationDate) {
				processStepProps.push(
					getSortableStep({
						heading: 'Income redetermined',
						body: `on ${dateFormatter(determination.determinationDate, false)}`,
						stepDate: determination.determinationDate,
					})
				);
			}
		});
	}

	return processStepProps;
}

/**
 * Generates SortableProcessStepProps array for a given array of C4KCertificates
 * @param c4KCertificates
 */
export function c4kCertificateStepProps(
	c4KCertificates: DeepNonUndefineable<C4KCertificate[]> | null
) {
	var processStepProps: SortableProcessStepProps[] = [];

	if (c4KCertificates) {
		c4KCertificates.sort(c4kCertificateSorter).forEach((c4kCertificate, idx) => {
			if (c4kCertificate.startDate) {
				processStepProps.push(
					getSortableStep({
						heading:
							idx === 0 ? 'Care 4 Kids certificate added' : 'Care 4 Kids certificate renewed',
						body: `on ${dateFormatter(c4kCertificate.startDate, false)}`,
						stepDate: c4kCertificate.startDate,
					})
				);
			}
		});
	}

	return processStepProps;
}

/**
 * Sorts SortableProcessStepProp objects inversely by sortDate, using sortWeight as a tie breaker
 * @param a
 * @param b
 */
export function enrollmentHistoryPropsSorter(
	a: SortableProcessStepProps,
	b: SortableProcessStepProps
) {
	var dateSorterRes = dateSorter(a.sortDate, b.sortDate, true);

	if (dateSorterRes !== 0) return dateSorterRes;

	if (a.sortWeight < b.sortWeight) return 1;
	if (a.sortWeight > b.sortWeight) return -1;
	return 0;
}

/**
 * Returns a SortableProcessStepProps object composed from the given parameters,
 * using the stepDate as sortDate, and to get correct value for isNew
 * @param heading
 * @param body
 * @param stepDate
 * @param stepWeight
 */
export function getSortableStep(params: {
	heading: string;
	body: string;
	stepDate: Date;
	stepWeight?: number;
}): SortableProcessStepProps {
	return {
		heading: params.heading,
		body: params.body,
		sortDate: params.stepDate,
		sortWeight: params.stepWeight ? params.stepWeight : 0,
		isNew: stepIsNew(params.stepDate),
	};
}

/**
 * Returns true is date is less than 1 week old, else false
 * @param stepDate
 */
export function stepIsNew(stepDate: Date) {
	return moment().add(-1, 'week') < moment(stepDate);
}
