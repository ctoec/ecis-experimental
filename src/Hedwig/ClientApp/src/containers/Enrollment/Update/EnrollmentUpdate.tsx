import React, { useContext, useState, useEffect } from 'react';
import { Section } from '../enrollmentTypes';
import { History } from 'history';
import ChildInfo from '../_sections/ChildInfo';
import FamilyInfo from '../_sections/FamilyInfo';
import FamilyIncome from '../_sections/FamilyIncome';
import EnrollmentFunding from '../_sections/EnrollmentFunding';
import PageNotFound from '../../PageNotFound/PageNotFound';
import UserContext from '../../../contexts/User/UserContext';
import {
	Enrollment,
	ApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGetRequest,
	ApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdPutRequest,
} from '../../../generated';
import { validatePermissions, getIdForUser } from '../../../utils/models';
import CommonContainer from '../../CommonContainer';
import { nameFormatter, editSaveFailAlert } from '../../../utils/stringFormatters';
import { ErrorBoundary } from '../../../components';
import useApi from '../../../hooks/useApi';
import { DeepNonUndefineable } from '../../../utils/types';

type EnrollmentUpdateParams = {
	history: History;
	match: {
		params: {
			siteId: number;
			enrollmentId: number;
			sectionId: string;
		};
	};
};

const sections: { [key: string]: Section } = {
	'child-information': ChildInfo,
	'family-information': FamilyInfo,
	'family-income': FamilyIncome,
	'enrollment-funding': EnrollmentFunding,
};

/**
 * React component for updating an enrollment. Renders a section
 * update form component.
 *
 * @param props Props with location.
 */
export default function EnrollmentUpdate({
	history,
	match: {
		params: { siteId, enrollmentId, sectionId },
	},
}: EnrollmentUpdateParams) {
	const section = sections[sectionId];
	const { user } = useContext(UserContext);
	const [attemptingSave, setAttemptingSave] = useState(false);
	const [enrollment, updateEnrollment] = useState<Enrollment | null>(null);
	// Get enrollment by id
	const params: ApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGetRequest = {
		id: enrollmentId ? enrollmentId : 0,
		orgId: getIdForUser(user, 'org'),
		siteId: validatePermissions(user, 'site', siteId) ? siteId : 0,
		include: ['child', 'family', 'determinations', 'fundings'],
	};
	const { loading, error, data: _enrollment } = useApi(
		(api) => api.apiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGet(params),
		{ skip: !user }
	);
	useEffect(() => {
		updateEnrollment(_enrollment);
	}, [_enrollment]);

	const saveParams: ApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdPutRequest = {
		id: enrollment ? enrollment.id : 0,
		orgId: getIdForUser(user, 'org'),
		siteId: validatePermissions(user, 'site', siteId) ? siteId : 0,
		enrollment: enrollment || undefined,
	};
	const { loading: saveLoading, error: saveError, data: saveData } = useApi<Enrollment>(
		(api) => api.apiOrganizationsOrgIdSitesSiteIdEnrollmentsIdPut(saveParams),
		{ skip: !attemptingSave || !user, callback: () => setAttemptingSave(false) }
	);
	useEffect(() => {
		// If the request went through, then do the next steps
		if (!saveData && !saveError) {
			return;
		}
		// Set the new error regardless of whether there is one
		if (saveData && !saveError) {
			updateEnrollment(saveData);
		} else {
		}
	}, [saveData, saveError]);

	if (!section) {
		return <PageNotFound />;
	}

	if (loading || !enrollment) {
		return <div className="EnrollmentUpdate"></div>;
	}

	// If update hasn't been implemented, default to edit
	if (!section.UpdateForm) {
		history.push(`/roster/sites/${siteId}/enrollments/${enrollment.id}/edit/${sectionId}`);
		return <></>;
	}

	return (
		<CommonContainer
			directionalLinkProps={{
				direction: 'left',
				to: `/roster/sites/${siteId}/enrollments/${enrollment.id}/`,
				text: `Back to summary`,
			}}
		>
			<div className="grid-container">
				<h1 className="margin-y-4">
					Update {section.name.toLowerCase()} for {nameFormatter(enrollment.child)}
				</h1>
				<ErrorBoundary alertProps={editSaveFailAlert}>
					<section.UpdateForm
						siteId={siteId}
						updateEnrollment={(enrollment) => {
							updateEnrollment(enrollment);
							setAttemptingSave(true);
						}}
						loading={saveLoading || attemptingSave}
						error={error || saveError}
						enrollment={enrollment}
						success={saveData && !saveError}
					/>
				</ErrorBoundary>
			</div>
		</CommonContainer>
	);
}
