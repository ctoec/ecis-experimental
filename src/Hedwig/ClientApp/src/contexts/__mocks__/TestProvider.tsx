import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory, History } from 'history';
import AuthenticationContext from '../Authentication/AuthenticationContext';
import UserContext from '../User/UserContext';
import ReportingPeriodContext from '../ReportingPeriod/ReportingPeriodContext';
import { User, ReportingPeriod } from '../../generated';
import { cdcReportingPeriods as _cdcReportingPeriods, user as _user } from '../../tests/data';

type TestProviderProps = {
	user?: User;
	cdcReportingPeriods?: ReportingPeriod[];
	history?: History<any>;
};

const TestProvider: React.FC<TestProviderProps> = ({
	children,
	user = _user,
	cdcReportingPeriods = _cdcReportingPeriods,
	history = createMemoryHistory(),
}) => {
	return (
		<AuthenticationContext.Provider
			value={{ accessToken: '', withFreshToken: jest.fn(), loading: false }}
		>
			<UserContext.Provider value={{ user, loading: false }}>
				<ReportingPeriodContext.Provider value={{ cdcReportingPeriods }}>
					<Router history={history}>{children}</Router>
				</ReportingPeriodContext.Provider>
			</UserContext.Provider>
		</AuthenticationContext.Provider>
	);
};

export default TestProvider;
