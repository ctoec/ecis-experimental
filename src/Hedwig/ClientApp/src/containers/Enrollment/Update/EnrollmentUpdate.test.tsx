// Variables used in jest mockes -- must start with `mock`
import { mockAllFakeEnrollments, mockSite } from '../../../tests/data';
import mockUseApi, {
	mockApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGet,
	mockApiOrganizationsOrgIdSitesIdGet,
} from '../../../hooks/__mocks__/useApi';

// Jest mocks must occur before later imports
jest.mock('../../../hooks/useApi', () =>
	mockUseApi({
		apiOrganizationsOrgIdSitesIdGet: mockApiOrganizationsOrgIdSitesIdGet(mockSite),
		apiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGet: mockApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGet(
			mockAllFakeEnrollments
		),
	})
);

import React from 'react';
import { render } from '@testing-library/react';
import TestProvider from '../../../contexts/__mocks__/TestProvider';
import { mockCompleteEnrollment, mockEnrollmentMissingBirthCertId } from '../../../tests/data';
import { accessibilityTestHelper } from '../../../tests/helpers';
import EnrollmentUpdate from './EnrollmentUpdate';
import { createMemoryHistory } from 'history';

afterAll(() => {
	jest.resetModules();
});

describe('EnrollmentUpdate', () => {
	it('matches snapshot', () => {
		const history = createMemoryHistory();
		const { asFragment } = render(
			<TestProvider history={history}>
				<EnrollmentUpdate
					history={history}
					match={{
						params: {
							siteId: mockCompleteEnrollment.siteId,
							enrollmentId: mockCompleteEnrollment.id,
							sectionId: 'family-income',
						},
					}}
				/>
			</TestProvider>
		);
		expect(asFragment()).toMatchSnapshot();
	});

	const history = createMemoryHistory();
	accessibilityTestHelper(
		<TestProvider history={history}>
			<EnrollmentUpdate
				history={history}
				match={{
					params: {
						siteId: mockEnrollmentMissingBirthCertId.siteId,
						enrollmentId: mockEnrollmentMissingBirthCertId.id,
						sectionId: 'family-income',
					},
				}}
			/>
		</TestProvider>
	);
});
