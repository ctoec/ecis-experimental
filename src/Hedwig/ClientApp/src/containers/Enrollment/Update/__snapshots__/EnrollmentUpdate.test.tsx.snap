// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnrollmentUpdate matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid-container"
  >
    <a
      class="oec-directional-link"
      href="/roster/sites/1/enrollments/1/"
    >
      <svg
        class="oec-directional-link__arrow oec-directional-link__arrow--direction-left oec-directional-link__arrow--side-left"
      >
        arrowRight.svg
      </svg>
      Back to summary
    </a>
  </div>
  <div
    class="grid-container"
  >
    <h1
      class="margin-y-4"
    >
      Update family income determination for Lily Luna Potter
    </h1>
    <h2
      class="margin-bottom-1"
    >
      Family income determination
    </h2>
    <div
      class="display-flex align-center"
    >
      <h3>
        Current income determination
      </h3>
         
      <button
        class="usa-button usa-button--unstyled"
        type="button"
      >
        Add new income determination
      </button>
    </div>
    <div>
      <div
        class="oec-card"
      >
        <div
          class="oec-card-cell"
        >
          <div
            class="usa-grid"
          >
            <div
              class="grid-row"
            >
              <div
                class="grid-col"
              >
                <p>
                  Income not disclosed
                </p>
              </div>
              <div
                class="grid-col-1 flex-align-self--center"
              >
                <button
                  class="usa-button usa-button--unstyled"
                  type="button"
                >
                  Edit
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class=""
        />
        <div
          class="oec-card-cell oec-card-expansion"
          hidden=""
        >
          <form
            autocomplete="off"
            class="FamilyIncomeForm"
            novalidate=""
          >
            <div
              class="usa-form"
            >
              <p
                class="text-bold font-sans-lg margin-top-2"
              >
                Edit family income determination
              </p>
              <div
                class="usa-form-group margin-top-3"
              >
                <div
                  class="margin-top-3"
                />
                <div
                  class="usa-checkbox undefined"
                >
                  <input
                    checked=""
                    class="usa-checkbox__input"
                    id="family-income-disclosed-0-familyIncomeNotDisclosed"
                    name="child.family.determinations.0.notDisclosed"
                    type="checkbox"
                    value="familyIncomeNotDisclosed"
                  />
                  <label
                    class="usa-checkbox__label"
                    for="family-income-disclosed-0-familyIncomeNotDisclosed"
                  >
                    Family income not disclosed
                  </label>
                </div>
              </div>
              <div
                class="display-flex"
              >
                <div
                  class="usa-form"
                >
                  <button
                    class="usa-button usa-button--outline"
                    type="button"
                  >
                    Cancel
                  </button>
                  <input
                    class="usa-button"
                    type="submit"
                    value="Save"
                  />
                </div>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
